2019/pkg/intcode/intcode.go

```diff
@@ -0,0 +1,154 @@
+package intcode
+
+import (
+	"errors"
+	"fmt"
+)
+
+type opCode int
+
+const (
+	opCodeAdd       opCode = 1
+	opCodeMultiply  opCode = 2
+	opCodeRead      opCode = 3
+	opCodeWrite     opCode = 4
+	opCodeTerminate opCode = 99
+	opCodeUnknown   opCode = -1
+)
+
+var intToOpCode = map[int]opCode{
```

Comment:
I'm not quite sure why you need this extra table if the values already map 1 to 1.

Reply:
I wanted a way to quickly look up if an int was a valid opcode (ex: [here](https://github.com/JessicaGreben/adventofcode/pull/3/files#diff-980e1de91dfd105039ade7de39fad36824c279514bab9aafd4dc5d098e0cc36eR120)).  Without this table, I was doing logic like this to verify that the opcode was a valid which would need to be updated each time a new opcode is added:
```
if (opcode >= 5 && opcode <= 8)  || opcode == 0 {
    // not valid opcode
}
```

Reply:
Ah, in that case I would use something like:
```
type opCode byte

const (
	opCodeAdd       opCode = 1
	opCodeMultiply  opCode = 2
	opCodeRead      opCode = 3
	opCodeWrite     opCode = 4
	opCodeTerminate opCode = 99
	opCodeUnknown   opCode = -1
)

var isValidOp = [256]bool{
	opCodeAdd: true,
	opCodeMultiply: true,
	opCodeRead: true,
	opCodeWrite: true,
	opCodeTerminate: true,
}
```

Or alternatively:
```
var isValid = map[opCode]bool{
```

Reply:
Interesting, I haven't seen this key-index syntax before that you used in  `isValidOp` array. Had to run that example thru go playground and ended up over [here](https://stackoverflow.com/a/36302875/4410404).  I like the map `isValid` a little more since it seems like more common code and readable. Do you agree?  What do you think are the pros/cons of the array `isValidOp` vs map `isValid`?

Reply:
The array version is faster, however it wouldn't be feasible `int32` (or larger data types) due to the memory requirements. `map` handles out-of-bounds values out-of-the-box, so if you use `if isValidOp[-1200] {` directly and it'll return the default value `false`.